From 97b2fa99f5da6dc2187ee3596cd03f89d8f6ab3f Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Wed, 6 Aug 2014 03:50:27 -0400
Subject: [pci] Don't enable decoding if BAR hasn't been assigned an address

Message-id: <20140806035027.1438.99009.stgit@gir.stowe>
Patchwork-id: 87468
O-Subject: [RHEL7.1 PATCH 59/69] PCI: Don't enable decoding if BAR hasn't been assigned an address
Bugzilla: 1110896
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110896
Upstream Status: 3cedcc3621289d41bd21c5dbe0b886d57c83a1ea

commit 3cedcc3621289d41bd21c5dbe0b886d57c83a1ea
Author: Bjorn Helgaas <bhelgaas@google.com>
Date:   Wed Feb 26 11:26:00 2014 -0700

    PCI: Don't enable decoding if BAR hasn't been assigned an address

    Don't enable memory or I/O decoding if we haven't assigned or claimed the
    BAR's resource.

    If we enable decoding for a BAR that hasn't been assigned an address, we'll
    likely cause bus conflicts.  This declines to enable decoding for resources
    with IORESOURCE_UNSET.

    Note that drivers can use pci_enable_device_io() or pci_enable_device_mem()
    if they only care about specific types of BARs.  In that case, we don't
    bother checking whether the corresponding resources are assigned or
    claimed.

    Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/pci/setup-res.c b/drivers/pci/setup-res.c
index 6e44313..7eed671 100644
--- a/drivers/pci/setup-res.c
+++ b/drivers/pci/setup-res.c
@@ -343,9 +343,15 @@ int pci_enable_resources(struct pci_dev *dev, int mask)
     (!(r->flags & IORESOURCE_ROM_ENABLE)))
    continue;
 
+  if (r->flags & IORESOURCE_UNSET) {
+   dev_err(&dev->dev, "can't enable device: BAR %d %pR not assigned\n",
+    i, r);
+   return -EINVAL;
+  }
+
   if (!r->parent) {
-   dev_err(&dev->dev, "device not available "
-    "(can't reserve %pR)\n", r);
+   dev_err(&dev->dev, "can't enable device: BAR %d %pR not claimed\n",
+    i, r);
    return -EINVAL;
   }
 
-- 
1.7.1